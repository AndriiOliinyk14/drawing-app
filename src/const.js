export const LS_KEY = "images";

export const TOOLS = {
  eraser: "ERASER",
  brush: "BRUSH",
};

export const INITIAL_PARAMETERS = {
  width: 600,
  height: 500,
  isMouseDown: false,
  size: 10,
  color: "red",
  background: "#ffffff",
  lineCap: "round",
  lineJoin: "round",
  tool: TOOLS.brush,
};
