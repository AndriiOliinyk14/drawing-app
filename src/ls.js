import { LS_KEY } from "./const.js";

export const saveProject = (name, item) => {
  const prevImages = localStorage.getItem(LS_KEY)
    ? JSON.parse(localStorage.getItem(LS_KEY))
    : {};

  if (prevImages && Object.keys(prevImages).includes(name)) {
    alert("This name already exists, please enter another name");
    return false;
  }

  localStorage.setItem(LS_KEY, JSON.stringify({ ...prevImages, [name]: item }));

  return true;
};

export const loadProjects = () => {
  return localStorage.getItem(LS_KEY)
    ? JSON.parse(localStorage.getItem(LS_KEY))
    : {};
};

export const loadProject = (item) => {
  const images = loadProjects();

  if (images[item]) {
    return images[item];
  } else {
    alert("Project not found");
  }
};

export const removeAllProjects = () => {
  localStorage.removeItem(LS_KEY);
};
