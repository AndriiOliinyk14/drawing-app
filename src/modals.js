import { saveProject, loadProjects, removeAllProjects } from "./ls.js";

const modal = document.getElementById("modal");
const modalTitle = document.getElementById("modal-title");
const modalContent = document.getElementById("modal-content");
const modalCloser = document.getElementById("modal-closer");

export const handleChangeModalTitle = (value) => {
  modalTitle.innerText = value;
};

export const handleOpenModal = (value) => {
  modal.style.visibility = "visible";
};

export const handleCloseModal = () => {
  modal.style.visibility = "hidden";
};

const handleClearModalContent = () => {
  modalContent.replaceChildren();
};

export const handleSave = (data) => {
  handleClearModalContent();
  const inputEl = document.createElement("input");
  const btnEl = document.createElement("button");

  inputEl.value = "Untitled";
  inputEl.className = "modal-input";

  btnEl.className = "modal-btn";
  btnEl.innerText = "Save";

  handleOpenModal();
  handleChangeModalTitle("Save project");

  const content = [inputEl, btnEl];

  modalContent.replaceChildren(...content);

  btnEl.addEventListener("click", () => {
    const name = inputEl.value.trim();

    if (saveProject(name, data)) {
      handleCloseModal();
    }
  });
};

export const handleLoad = (callback) => {
  handleClearModalContent();
  handleOpenModal();
  handleChangeModalTitle("Projects");

  let projects = loadProjects();

  const listContainer = document.createElement("ul");
  const removeAllBtn = document.createElement("button");

  removeAllBtn.innerText = "Remove all projects";
  removeAllBtn.className = "modal-btn";
  listContainer.className = "modal-load-list";

  if (!Object.keys(projects).length) {
    handleChangeModalTitle("Not found projects");
    return;
  }

  const renderList = () => {
    const items = Object.keys(projects).map((key) => {
      const li = document.createElement("li");
      li.className = "modal-load-item";
      const btn = document.createElement("button");

      btn.addEventListener("click", () => {
        callback(projects[key]);
        handleCloseModal();
      });

      btn.innerText = key;
      li.replaceChildren(btn);
      return li;
    });

    listContainer.replaceChildren(...items);
  };

  removeAllBtn.addEventListener("click", () => {
    if (window.confirm("You really want to remove all projects")) {
      removeAllProjects();
      alert("All projects have been deleted");
      handleCloseModal();
    }
  });

  renderList();

  modalContent.append(...[listContainer, removeAllBtn]);
};

document.addEventListener("DOMContentLoaded", () => {
  modalCloser.addEventListener("click", handleCloseModal);
});
