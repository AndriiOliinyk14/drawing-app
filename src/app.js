import { TOOLS, INITIAL_PARAMETERS } from "./const.js";
import { handleLoad, handleSave } from "./modals.js";

const canvas = document.getElementById("canvas");

const saveEl = document.getElementById("save");
const loadEl = document.getElementById("load");
const exportEl = document.getElementById("export");
const eraserEl = document.getElementById("eraser");
const brushEl = document.getElementById("brush");
const clearEl = document.getElementById("clear");
const sizeEl = document.getElementById("size");
const colorEl = document.getElementById("color");
const widthEl = document.getElementById("width");
const heightEl = document.getElementById("height");

const ctx = canvas.getContext("2d");

widthEl.value = INITIAL_PARAMETERS.width;
heightEl.value = INITIAL_PARAMETERS.height;
sizeEl.value = INITIAL_PARAMETERS.size;
colorEl.value = INITIAL_PARAMETERS.color;
brushEl.classList.add("selected");

export const app = {
  linesArray: [],

  parameters: {
    width: widthEl.value,
    height: heightEl.value,
    isMouseDown: false,
    size: sizeEl.value,
    color: colorEl.value,
    background: "#ffffff",
    lineCap: "round",
    lineJoin: "round",
    tool: TOOLS.brush,
  },

  changeTool(tool) {
    this.parameters.tool = tool;

    const handleRemoveSelectedClass = (el) => {
      el.classList.remove("selected");
    };

    if (tool === TOOLS.brush) {
      handleRemoveSelectedClass(eraserEl);
      brushEl.classList.add("selected");
    } else {
      handleRemoveSelectedClass(brushEl);
      eraserEl.classList.add("selected");
    }
  },

  getMousePos(canvas, evt) {
    const rect = canvas.getBoundingClientRect();
    return {
      x: evt.clientX - rect.left,
      y: evt.clientY - rect.top,
    };
  },

  createCanvas() {
    canvas.height = this.parameters.height;
    canvas.width = this.parameters.width;
    ctx.fillStyle = this.parameters.background;
    ctx.imageSmoothingEnabled = true;
    ctx.fillRect(0, 0, canvas.width, canvas.height);
  },

  redraw() {
    this.linesArray.forEach((line, idx) => {
      ctx.beginPath();
      ctx.moveTo(line.x, line.y);
      ctx.lineWidth = line.size;
      ctx.lineCap = this.parameters.lineCap;
      ctx.lineJoin = this.parameters.lineJoin;
      ctx.strokeStyle =
        line.tool === TOOLS.brush ? line.color : this.parameters.background;
      ctx.lineTo(this.linesArray[idx + 1]?.x, this.linesArray[idx + 1]?.y);
      ctx.stroke();
    });
  },

  resetParameters() {
    this.parameters = INITIAL_PARAMETERS;
  },

  clearAll() {
    this.linesArray = [];
    ctx.clearRect(0, 0, canvas.width, canvas.height);
  },

  onMouseDown(canvas, evt) {
    const mousePos = this.getMousePos(canvas, evt);
    this.parameters.isMouseDown = true;
    const currentPosition = this.getMousePos(canvas, evt);
    ctx.beginPath();
    ctx.moveTo(currentPosition.x, currentPosition.y);
    ctx.lineWidth = this.parameters.size;
    ctx.lineCap = this.parameters.lineCap;
    ctx.lineJoin = this.parameters.lineJoin;
    ctx.strokeStyle =
      this.parameters.tool === TOOLS.brush
        ? this.parameters.color
        : this.parameters.background;
  },

  onMouseMove(canvas, evt) {
    if (this.parameters.isMouseDown) {
      const currentPosition = this.getMousePos(canvas, evt);
      ctx.lineTo(currentPosition.x, currentPosition.y);
      ctx.stroke();
      this.store(
        currentPosition.x,
        currentPosition.y,
        this.parameters.size,
        this.parameters.color,
        this.parameters.tool
      );
    }
  },

  onMouseUp() {
    this.parameters.isMouseDown = false;
    this.store();
  },

  store(x, y, s, c, t) {
    const line = {
      x,
      y,
      size: s,
      color: c,
      tool: t,
    };

    this.linesArray.push(line);
  },

  save() {
    handleSave({
      lines: this.linesArray,
      size: { width: this.parameters.width, height: this.parameters.height },
    });
  },

  load() {
    const handleLoading = ({ lines, size }) => {
      this.parameters.width = size.width;
      this.parameters.height = size.height;
      this.clearAll();
      this.linesArray = lines;
      this.createCanvas();
      this.redraw();
    };

    handleLoad(handleLoading);
  },

  export() {
    const img = canvas.toDataURL("image/png");
    exportEl.download = "image.png";
    exportEl.href = img;
  },
};

canvas.addEventListener("mousedown", function (evt) {
  app.onMouseDown(canvas, evt);
});

canvas.addEventListener("mousemove", function (evt) {
  app.onMouseMove(canvas, evt);
});

canvas.addEventListener("mouseup", function () {
  app.onMouseUp();
});

document.addEventListener("mouseup", function () {
  app.onMouseUp();
});

eraserEl.addEventListener("click", function () {
  app.changeTool(TOOLS.eraser);
});

brushEl.addEventListener("click", function () {
  app.changeTool(TOOLS.brush);
});

clearEl.addEventListener("click", function () {
  app.clearAll();
});

saveEl.addEventListener("click", function () {
  app.save();
});

loadEl.addEventListener("click", function () {
  app.load();
});

exportEl.addEventListener("click", function () {
  app.export();
});

sizeEl.addEventListener("change", function (evt) {
  app.parameters.size = evt.target.value;
});

colorEl.addEventListener("change", function (evt) {
  app.parameters.color = evt.target.value;
});

widthEl.addEventListener("change", function (evt) {
  app.parameters.width = evt.target.value;
  app.createCanvas();
  app.redraw();
});

heightEl.addEventListener("change", function (evt) {
  app.parameters.height = evt.target.value;
  app.createCanvas();
  app.redraw();
});
