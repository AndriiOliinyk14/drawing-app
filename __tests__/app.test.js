/**
 * @jest-environment jsdom
 */

import fs from "fs";
import path from "path";
import { describe, expect, beforeEach } from "@jest/globals";
import { TOOLS, INITIAL_PARAMETERS } from "../src/const";

const html = fs.readFileSync(path.resolve(__dirname, "../index.html"), "utf8");

const paths = [
  { x: 30, y: 40, size: 10, color: "#E71818", tool: "BRUSH" },
  {
    x: undefined,
    y: undefined,
    size: undefined,
    color: undefined,
    tool: undefined,
  },
];

beforeAll(() => {
  // Set up the DOM for testing
  document.documentElement.innerHTML = html.toString();
});

beforeEach(() => {
  const { app } = require("../src/app");

  app.clearAll();
});

describe("application", () => {
  it("should change tool", () => {
    const { app } = require("../src/app");

    const brushEl = document.getElementById("brush");
    const eraserEl = document.getElementById("eraser");

    const brushMockFn = jest.fn();
    const eraserMockFn = jest.fn();

    brushEl.addEventListener("click", brushMockFn);
    brushEl.click();

    expect(brushMockFn).toBeCalled();
    expect(app.parameters.tool).toBe(TOOLS.brush);

    app.changeTool(TOOLS.eraser);

    eraserEl.addEventListener("click", eraserMockFn);
    eraserEl.click();

    expect(eraserMockFn).toBeCalled();
    expect(app.parameters.tool).toBe(TOOLS.eraser);
  });

  it("should get mouse position", () => {
    const { app } = require("../src/app");

    const canvas = document.getElementById("canvas");

    const eventData = {
      clientX: 10,
      clientY: 10,
    };

    expect(app.getMousePos(canvas, eventData)).toEqual({ x: 10, y: 10 });
  });

  it("should render canvas", () => {
    const { app } = require("../src/app");

    app.createCanvas();

    const canvas = document.getElementById("canvas");

    expect(canvas).toBeTruthy();
    expect(canvas.width).toEqual(INITIAL_PARAMETERS.width);
    expect(canvas.height).toEqual(INITIAL_PARAMETERS.height);
  });

  it("should reset to default parameters", () => {
    const { app } = require("../src/app");

    app.changeTool(TOOLS.eraser);

    expect(app.parameters.tool).toEqual(TOOLS.eraser);

    app.resetParameters();

    expect(app.parameters.tool).toEqual(TOOLS.brush);
    expect(app.parameters).toBe(INITIAL_PARAMETERS);
  });

  it("should draw", () => {
    const { app } = require("../src/app");

    app.createCanvas();

    const canvas = document.getElementById("canvas");
    const ctx = canvas.getContext("2d");

    const eventDataDown = {
      clientX: 10,
      clientY: 10,
    };

    const eventDataMove = {
      clientX: paths[0].x,
      clientY: paths[0].y,
    };

    app.onMouseDown(canvas, eventDataDown);
    app.onMouseMove(canvas, eventDataMove);
    app.onMouseUp();

    expect(paths).toStrictEqual(app.linesArray);
  });

  it("should clear all", () => {
    const { app } = require("../src/app");

    app.createCanvas();

    const canvas = document.getElementById("canvas");

    const eventDataDown = {
      clientX: 10,
      clientY: 10,
    };

    const eventDataMove = {
      clientX: paths[0].x,
      clientY: paths[0].y,
    };

    app.onMouseDown(canvas, eventDataDown);
    app.onMouseMove(canvas, eventDataMove);
    app.onMouseUp();

    expect(paths).toStrictEqual(app.linesArray);
    expect(app.linesArray).toHaveLength(paths.length);

    app.clearAll();

    expect(app.linesArray).toHaveLength(0);
  });

  it("should redraw", () => {
    const { app } = require("../src/app");

    app.createCanvas();

    const canvas = document.getElementById("canvas");
    const ctx = canvas.getContext("2d");

    app.linesArray = paths;
    app.redraw();

    expect(ctx.moveTo).toBeCalledWith(paths[0].x, paths[0].y);
    expect(app.linesArray).toStrictEqual(paths);
  });

  it("should call save function", () => {
    const { app } = require("../src/app");
    const mockFc = jest.spyOn(app, "save");

    const saveEl = document.getElementById("save");

    saveEl.click();

    expect(mockFc).toBeCalledTimes(1);
  });

  it("should call load function", () => {
    const { app } = require("../src/app");
    const mockFc = jest.spyOn(app, "load");

    const loadEl = document.getElementById("load");

    loadEl.click();

    expect(mockFc).toBeCalledTimes(1);
  });

  it("should call export function", () => {
    const { app } = require("../src/app");
    const mockFc = jest.spyOn(app, "export");

    const exportEl = document.getElementById("export");

    exportEl.click();

    expect(mockFc).toBeCalledTimes(1);
  });
});

describe("DOM Event", () => {
  it("should listening mousedown event", () => {
    const mockFn = jest.fn();

    const canvas = document.getElementById("canvas");

    canvas.addEventListener("mousedown", mockFn);

    const event = new MouseEvent("mousedown", {
      bubbles: true,
      cancelable: true,
      view: window,
    });

    canvas.dispatchEvent(event);

    expect(mockFn).toHaveBeenCalled();
  });

  it("should listening mouseup of canvas event", () => {
    const mockFn = jest.fn();

    const canvas = document.getElementById("canvas");

    canvas.addEventListener("mouseup", mockFn);

    const event = new MouseEvent("mouseup", {
      bubbles: true,
      cancelable: true,
      view: window,
    });

    canvas.dispatchEvent(event);

    expect(mockFn).toHaveBeenCalled();
  });

  it("should listening mouseup of document event", () => {
    const mockFn = jest.fn();

    document.addEventListener("mouseup", mockFn);

    const event = new MouseEvent("mouseup", {
      bubbles: true,
      cancelable: true,
      view: window,
    });

    document.dispatchEvent(event);

    expect(mockFn).toHaveBeenCalled();
  });

  it("should listening mousemove event", () => {
    const mockFn = jest.fn();

    const canvas = document.getElementById("canvas");

    canvas.addEventListener("mousemove", mockFn);

    const event = new MouseEvent("mousemove", {
      bubbles: true,
      cancelable: true,
      view: window,
      clientX: 100,
      clientY: 200,
    });




    canvas.dispatchEvent(event);

    expect(mockFn).toHaveBeenCalled();
  });

  it("should be clicked on clear button", () => {
    const mockFn = jest.fn();

    const button = document.getElementById("clear");

    button.addEventListener("click", mockFn);

    button.click();

    expect(mockFn).toHaveBeenCalled();
  });
});
