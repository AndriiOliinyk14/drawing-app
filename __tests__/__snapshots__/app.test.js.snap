// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application should draw 1`] = `
CanvasRenderingContext2D {
  "_canvas": <canvas
    class="canvas"
    height="500"
    id="canvas"
    width="600"
  />,
  "_clipIndex": 1,
  "_clipStack": [
    [],
  ],
  "_directionStack": [
    "inherit",
  ],
  "_drawCalls": [
    {
      "props": {
        "height": 150,
        "width": 300,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 150,
        "width": 300,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 150,
        "width": 300,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillRect",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillRect",
    },
    {
      "props": {
        "path": [
          {
            "props": {},
            "transform": [
              1,
              0,
              0,
              1,
              0,
              0,
            ],
            "type": "beginPath",
          },
          {
            "props": {
              "x": 10,
              "y": 10,
            },
            "transform": [
              1,
              0,
              0,
              1,
              0,
              0,
            ],
            "type": "moveTo",
          },
          {
            "props": {
              "x": 30,
              "y": 40,
            },
            "transform": [
              1,
              0,
              0,
              1,
              0,
              0,
            ],
            "type": "lineTo",
          },
        ],
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "stroke",
    },
  ],
  "_events": [
    {
      "props": {
        "height": 150,
        "width": 300,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 150,
        "width": 300,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 150,
        "width": 300,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "value": "#ffffff",
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillStyle",
    },
    {
      "props": {
        "value": true,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "imageSmoothingEnabled",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillRect",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "clearRect",
    },
    {
      "props": {
        "value": "#ffffff",
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillStyle",
    },
    {
      "props": {
        "value": true,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "imageSmoothingEnabled",
    },
    {
      "props": {
        "height": 500,
        "width": 600,
        "x": 0,
        "y": 0,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "fillRect",
    },
    {
      "props": {},
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "beginPath",
    },
    {
      "props": {
        "x": 10,
        "y": 10,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "moveTo",
    },
    {
      "props": {
        "value": 10,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "lineWidth",
    },
    {
      "props": {
        "value": "round",
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "lineCap",
    },
    {
      "props": {
        "value": "round",
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "lineJoin",
    },
    {
      "props": {
        "value": "#e71818",
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "strokeStyle",
    },
    {
      "props": {
        "x": 30,
        "y": 40,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "lineTo",
    },
    {
      "props": {
        "path": [
          {
            "props": {},
            "transform": [
              1,
              0,
              0,
              1,
              0,
              0,
            ],
            "type": "beginPath",
          },
          {
            "props": {
              "x": 10,
              "y": 10,
            },
            "transform": [
              1,
              0,
              0,
              1,
              0,
              0,
            ],
            "type": "moveTo",
          },
          {
            "props": {
              "x": 30,
              "y": 40,
            },
            "transform": [
              1,
              0,
              0,
              1,
              0,
              0,
            ],
            "type": "lineTo",
          },
        ],
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "stroke",
    },
  ],
  "_fillStyleStack": [
    "#ffffff",
  ],
  "_filterStack": [
    "none",
  ],
  "_fontStack": [
    "10px sans-serif",
  ],
  "_globalAlphaStack": [
    1,
  ],
  "_globalCompositeOperationStack": [
    "source-over",
  ],
  "_imageSmoothingEnabledStack": [
    true,
  ],
  "_imageSmoothingQualityStack": [
    "low",
  ],
  "_lineCapStack": [
    "round",
  ],
  "_lineDashOffsetStack": [
    0,
  ],
  "_lineDashStack": [
    [],
  ],
  "_lineJoinStack": [
    "round",
  ],
  "_lineWidthStack": [
    10,
  ],
  "_miterLimitStack": [
    10,
  ],
  "_path": [
    {
      "props": {},
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "beginPath",
    },
    {
      "props": {
        "x": 10,
        "y": 10,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "moveTo",
    },
    {
      "props": {
        "x": 30,
        "y": 40,
      },
      "transform": [
        1,
        0,
        0,
        1,
        0,
        0,
      ],
      "type": "lineTo",
    },
  ],
  "_shadowBlurStack": [
    0,
  ],
  "_shadowColorStack": [
    "rgba(0, 0, 0, 0)",
  ],
  "_shadowOffsetXStack": [
    0,
  ],
  "_shadowOffsetYStack": [
    0,
  ],
  "_stackIndex": 0,
  "_strokeStyleStack": [
    "#e71818",
  ],
  "_textAlignStack": [
    "start",
  ],
  "_textBaselineStack": [
    "alphabetic",
  ],
  "_transformStack": [
    [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
  ],
  "addHitRegion": [MockFunction],
  "arc": [MockFunction],
  "arcTo": [MockFunction],
  "beginPath": [MockFunction] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "bezierCurveTo": [MockFunction],
  "clearHitRegions": [MockFunction],
  "clearRect": [MockFunction] {
    "calls": [
      [
        0,
        0,
        600,
        500,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "clip": [MockFunction],
  "closePath": [MockFunction],
  "createImageData": [MockFunction],
  "createLinearGradient": [MockFunction],
  "createPattern": [MockFunction],
  "createRadialGradient": [MockFunction],
  "drawFocusIfNeeded": [MockFunction],
  "drawImage": [MockFunction],
  "ellipse": [MockFunction],
  "fill": [MockFunction],
  "fillRect": [MockFunction] {
    "calls": [
      [
        0,
        0,
        600,
        500,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "fillText": [MockFunction],
  "getImageData": [MockFunction],
  "getLineDash": [MockFunction],
  "getTransform": [MockFunction],
  "isPointInPath": [MockFunction],
  "isPointInStroke": [MockFunction],
  "lineTo": [MockFunction] {
    "calls": [
      [
        30,
        40,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "measureText": [MockFunction],
  "moveTo": [MockFunction] {
    "calls": [
      [
        10,
        10,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "putImageData": [MockFunction],
  "quadraticCurveTo": [MockFunction],
  "rect": [MockFunction],
  "removeHitRegion": [MockFunction],
  "resetTransform": [MockFunction],
  "restore": [MockFunction],
  "rotate": [MockFunction],
  "save": [MockFunction],
  "scale": [MockFunction],
  "scrollPathIntoView": [MockFunction],
  "setLineDash": [MockFunction],
  "setTransform": [MockFunction],
  "stroke": [MockFunction] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "strokeRect": [MockFunction],
  "strokeText": [MockFunction],
  "transform": [MockFunction],
  "translate": [MockFunction],
}
`;
